homeassistant/helpers/state.py

```diff
@@ -132,6 +89,109 @@ def get_changed_since(states: Iterable[State],
             if state.last_updated >= utc_point_in_time]
 
 
+@bind_hass
+def async_map_state_to_calls(hass: HomeAssistantType,
+                             domain: str,
+                             state: State,
+                             service_schemas: Iterable[Tuple[str, vol.Schema]],
+                             ) -> List[ServiceCall]:
+        """
+        Map state object into service calls to reproduce the given state.
+
+        Function is provided with a schema for each potential service. This
+        schema is extended with a state entry for services wich take no
+        parameter and modify state like  turn_on/off and such.
+        """
+        calls = []
+
+        # expanded dict containing entity id and state
+        attributes = dict(state.attributes)
+        attributes[ATTR_ENTITY_ID] = state.entity_id
+        attributes[ATTR_STATE] = state.state
+
+        for schema, service in service_schemas:
+            try:
+                validated = schema(attributes)
+            except vol.error.Invalid as e:
+                _LOGGER.debug("Incomplete attribute "
+                              "list to call {}:{}".format(service, e))
+                continue
+
+            # state is only syntetic for validation
+            validated.pop(ATTR_STATE, None)
+
+            # consume attributes taken by service
+            # this avoid multiple service calls
+            # for same attribute if schema would
+            # allow it
+            for key in validated:
+                attributes.pop(key, None)
```

Comment:
I'm still hesitant that we should handle this case in the helper. It makes the helper more unpredictable, regarding what service calls it will return.

Reply:
I cant allow both set operation mode and set temperature service for climate if not. Well i can allow it, but well get double calls.

I think it should be like this, but I can live with dropping it so I'll do that.

Reply:
Can we handle the climate case in the climate component instead? Maybe use a custom validation schema?

Reply:
Well, the schema must also map to a valid service call. Ie all attributes needed for the service and no more.

But we could use a custom schema for the set temperature service skipping operation mode since that is optional in the service. But given the climate architecture redesign peneing. I'm not sure it's worth it.

Reply:
It's totally fine to call multiple services to recreate a single scene state.

Reply:
The issue is that we can end up calling multiple services for a single attribute. Should normally not cause a problem as long as both services assume the same attribute have the same meaning and it's not some sort of toggling feature.